runtime/binding-http/src/main/java/io/aklivity/zilla/runtime/binding/http/internal/stream/HttpClientFactory.java

```diff
@@ -856,103 +859,148 @@ int decode(
-                countRequestsRejected.getAsLong();
+            return newStream;
+        }
 
-                // ignore DATA, FLUSH, END, ABORT
-                newStream = (t, b, i, l) -> {};
-            }
+        private MessageConsumer rejectWithStatusCode(
+                MessageConsumer sender,
+                BeginFW begin,
+                HttpBeginExFW beginEx)
+        {
+            MessageConsumer newStream;
+            // count all responses
+            countResponses.getAsLong();
+
+            final long sequence = begin.sequence();
+            final long acknowledge = begin.acknowledge();
+            final int maximum = begin.maximum();
+            final long routeId = begin.routeId();
+            final long initialId = begin.streamId();
+            final long traceId = begin.traceId();
+            final long authorization = begin.authorization();
+            final long replyId = supplyReplyId.applyAsLong(initialId);
+
+            doWindow(sender, routeId, initialId, 0, 0, 0, traceId, authorization, 0L, 0);
 
-            assert requests.size() <= maximumQueuedRequestsPerRoute;
+            doBegin(sender, routeId, replyId, sequence, acknowledge, maximum, supplyTraceId.getAsLong(), 0L, 0, beginEx);
+            doEnd(sender, routeId, replyId, sequence, acknowledge, maximum, supplyTraceId.getAsLong(), 0, EMPTY_OCTETS);
 
+            // count rejected requests (no connection or no space in the queue)
+            countRequestsRejected.getAsLong();
+
+            // ignore DATA, FLUSH, END, ABORT
+            newStream = (t, b, i, l) -> {};
             return newStream;
         }
 
         private void flushNext()
         {
-            if (!requests.isEmpty())
+            while (headerSlotOffset != headerSlotLimit)
             {
-                HttpClient client = supplyClient();
-
-                if (client != null)
+                final MutableDirectBuffer headerBuffer = bufferPool.buffer(headerSlot);
+                final long streamId = headerBuffer.getLong(headerSlotOffset);
+                final int length = headerBuffer.getInt(headerSlotOffset + Long.BYTES);
+                HttpExchange httpExchange = exchanges.get(streamId);
+                if (httpExchange != null)
                 {
-                    final HttpRequest request = requests.poll();
-                    assert request != null;
+                    httpExchange.doRequestBegin(headerBuffer, headerSlotOffset + Long.BYTES + Integer.BYTES, length);
+                }
+                headerSlotOffset += Long.BYTES + Integer.BYTES + length;
+                dequeues.getAsLong();
+            }
 
-                    dequeues.getAsLong();
+            headerSlotOffset = 0;
+            headerSlotLimit = 0;
+        }
 
-                    request.doFlushBegin(client);
-                }
+        private void cleanupHeaderSlotIfNecessary()
+        {
+            if (headerSlotOffset == headerSlotLimit && headerSlot != NO_SLOT)
```

Comment:
Logically we'd put the test for `headerSlot != NO_SLOT` first as that is gating for the other values to have meaning.